lua/terrortown/entities/items/item_ttt_radar/init.lua

```diff
@@ -66,36 +46,139 @@ local function ttt_radar_scan(ply, cmd, args)
 		table.Add(scan_ents, ents.FindByClass("ttt_decoy"))
 
 		for i = 1, #scan_ents do
-			local pl = scan_ents[i]
+			local ent = scan_ents[i]
 
-			if not IsValid(pl) or ply == pl or pl:IsPlayer() and (not pl:IsTerror() or pl:GetNWBool("disguised", false)) then continue end
+			if not IsValid(ent) or ply == ent or ent:IsPlayer() and (not ent:IsTerror() or ent:GetNWBool("disguised", false)) then continue end
 
-			local pos = pl:LocalToWorld(pl:OBBCenter())
+			local pos = ent:LocalToWorld(ent:OBBCenter())
 
 			-- Round off, easier to send and inaccuracy does not matter
 			pos.x = math.Round(pos.x)
 			pos.y = math.Round(pos.y)
 			pos.z = math.Round(pos.z)
 
-			local subrole = GetSubRoleForRadar(ply, pl)
-			local _tmp = {subrole = subrole, pos = pos}
-
-			targets[#targets + 1] = _tmp
+			targets[#targets + 1] = RADAR.CreateNewTarget(ply, pos, ent)
 		end
 	end
 
 	net.Start("TTT_Radar")
-	net.WriteUInt(#targets, 8)
+	net.WriteUInt(#targets, 16)
 
 	for i = 1, #targets do
 		local tgt = targets[i]
 
-		net.WriteUInt(tgt.subrole, ROLE_BITS)
 		net.WriteInt(tgt.pos.x, 32)
 		net.WriteInt(tgt.pos.y, 32)
 		net.WriteInt(tgt.pos.z, 32)
+
+		if tgt.subrole == -1 then
+			net.WriteBool(false)
+		else
+			net.WriteBool(true)
+			net.WriteUInt(tgt.subrole, ROLE_BITS)
+		end
+
+		if tgt.color then
+			net.WriteBool(true)
+			net.WriteColor(tgt.color)
+		else
+			net.WriteBool(false)
+		end
 	end
 
 	net.Send(ply)
 end
-concommand.Add("ttt_radar_scan", ttt_radar_scan)
+concommand.Add("ttt_radar_scan", TriggerRadarScan)
+
+local function GetDataForRadar(ply, ent)
+	local subrole = -1
+
+	if not IsValid(ent) then
+		subrole = -1
+	elseif not ent:IsPlayer() then
+		-- Decoys appear as innocents for players from other teams
+		if ent:GetNWString("decoy_owner_team", "none") ~= ply:GetTeam() then
+			subrole = ROLE_INNOCENT
+		end
+	else
+		local tmpRole = hook.Run("TTT2ModifyRadarRole", ply, ent)
+
+		if tmpRole then
+			subrole = tmpRole
+		else
+			subrole = (ent:IsInTeam(ply) or ent:GetSubRoleData().visibleForTraitors) and ent:GetSubRole() or ROLE_INNOCENT
+		end
+	end
+
+	return subrole
+end
+
+---
+-- Creates a new radar point
+-- @param Player ply The player that will see this radar point
+-- @param Vector pos The position of the radar point
+-- @param [opt]Entity ent The entity that is used for this radar point
+-- @param [opt]Color color A color for this radar point, this overwrites the normal color
+-- @realm server
+function RADAR.CreateNewTarget(ply, pos, ent, color)
```

Comment:
This name is quite misunderstandable. I would expect a function, which is creating the target point on radar, not one which returns the processed radar data table